### ISSUE TITLE
[Server] Logging settings in config/application.cr

### ISSUE THREAD

**Issue by docelic** [amberframework/amber#645]
### Description

There are inconsistencies present in how Amber logger settings are defined in the `Amber::Server.configure` block in `config/application.cr`.

Following the path of execution in `config/application.cr`, two relevant things happen:

1) First, "require "amber"" is called. At this point the environment and logger are initialized (this happens right then, not later).
2) Then, "Amber::Server.configure" block is called which is supposed to set (and override) any previous settings coming from the defaults and environment settings.

The logging-related options available in the `Amber::Server.configure` block are:

```
  # settings.logging.colorize = true
  # settings.logging.severity = debug
  # settings.logging.filter = %w(password confirm_password)
  # settings.logging.skip = %w()
  # settings.logging.context = %w(request headers cookies session params)
```

As can easily be seen from looking at these lines, these settings are applied on "logging", not on "logger". (`logger` is the actual existing logger, while `logging` is a template used to build and initialize new loggers.)

The inconsistencies stemming from that are as follows:

1) These config settings which are set on "logging" only affect future loggers; they do not affect the current logger. The only way to have them affect the "current" logger is to also uncomment the line near the end of "configure" block which says `settings.logger = Amber::Environment::Logger.new ...` to completely replace the current/already-initialized logger with a new one.

2) Some of the settings which exist on "logging" have non-standard names and values when compared to the Logger class. For example, in "logging", the severity is called "severity" while in Logger it is called "level". Also, severity (level) in "logging" is set as a lowercase string, while in Logger it is of type "Severity".

3) Some methods that exist on "logging" do not exist on "logger". (For example, it would be reasonable to support "logger.colorize = true/false", but this works only on "logging".)

4) Finally, even if a person does enable the `settings.logger = Amber::Environment::Logger.new` line near the end of the configure block to replace the current logger with the new one (which will be honoring the "logging" settings), this new logger will have a different message format than the first/original logger that was created by Amber.

### Steps to Reproduce

1. Create new app with `amber new testapp --deps; cd testapp`
2. Run `crystal src/testapp.cr` and make a couple HTTP requests to the app to get a feel for how the logger behaves
3. Edit `config/application.cr` and change any of the "logging" settings, then re-start the app
4. Note how nothing has changed in logger's behavior
5. Again edit `config/application.cr` to enable the `settings.logger = Amber::Environment::Logger.new` line near the end, then re-start the app
6. Note how the "logging" settings are now honored, but the message format is different than the one from built-in logger and this wasn't configured/requested by the user

**Expected behavior:**

Expected behavior would be that:

1) All logging settings are always standard and uniform, regardless of whether they are set on "logger" on "logging". This includes the 3 most obvious and mentioned differences -- severity/level name, accepted value type for severity/level, and support for #colorize = true/false on logger, but there may be others.

2) Settings should apply to both the current logger and future loggers.

3) If a person does create a new/separate logger by uncommenting the `settings.logger = ` line, its message format and all other non-changed settings are identical to Amber's default/first-created logger.

### Solution

As a first step, I would suggest unifying the interface that "logger" and "logging" accept. Primarily this means that:
1) Logging's "severity" should be called "level"
2) Logging's accepted value should be of type "Severity" instead of lowercase string
3) Logger should support method "colorize = true/false"

Then, we should either modify defaults so that calling Amber::Environment::Logger.new creates a logger with the identical settings, or that object is somehow cloned from existing Amber.settings.logger.

This would bring the whole behavior closer to a unified/identical interface, come with no surprises, be modifiable in runtime (as opposed to only for new logger objects), and generally work as one would expect.

**Comment by eliasjpr** [comment]
Go For it!

**Comment by eliasjpr** [comment]
Actually, the reason we have the Logging is to allow the logger settings to be configurable via the YAML. If you find a cleaner approach please feel free to propose a change

**Comment by eliasjpr** [comment]
@docelic please let me know if this still needs more work based on #683 changes

**Comment by docelic** [comment]
Hey @eliasjpr @epergo , yes, at least a little more work is needed. I see this:

If you run Amber with all default settings, logger messages look like this:

```
10:21:23 Server     | (INFO) Amber 0.6.7 serving application "Ss" at http://0.0.0.0:3000
10:21:23 Server     | (INFO) Server started in development.
10:21:23 Server     | (INFO) Startup Time 00:00:00.000061000
10:21:23 Watcher    | (INFO) Watching 7 files (browser reload)...
```

Then, if you open config/application.cr and uncomment the line near the end which says:

settings.logger = Amber::Environment::Logger.new(.....)

Then the messages become:

```
I, [2018-03-08 22:21:45 +01:00 #24193]  INFO --            |: Amber 0.6.7 serving application "Ss" at http://0.0.0.0:3000
I, [2018-03-08 22:21:45 +01:00 #24193]  INFO --            |: Server started in development.
I, [2018-03-08 22:21:45 +01:00 #24193]  INFO --            |: Startup Time 00:00:00.000066000
I, [2018-03-08 22:21:45 +01:00 #24193]  INFO -- Watcher    |: Watching 7 files (browser reload)...
```

I believe it would be nicer/more consistent if all loggers within the app inherited from the same logger definition, and that they are not different.

Feel free to close this issue if you do not think the difference shown above is a problem.

Thanks!

**Comment by epergo** [comment]
Hi @docelic, you are completely right! Definitely logs needs more work.

I think the general approach decided to manage logs is kind of correct, but it needs a couple of changes. What I have learned about this structure of Log settings/management is as follows:

* `Logger` (`amber/environment/logger.cr`) -> Wrapper of Crystal's `Logger` class, it only adds methods to log based on severity

* `Logging`  (`amber/environment/logging.cr`) -> Stores configuration of logs, either when specified in `your_environment.yml` or in `application.cr`, also stores logger's default settings in `DEFAULTS` constant.

* `LoggerBuilder` (`amber/environment/logger_builder.cr`) -> Factory class to create a new `Logger` with opinionated formatter defaults, using values obtained from `Logging`, this is the default Amber's logger formatter settings

So, whenever you want to instantiate the same type of Logger that Amber uses by default, you (as a user) won't call `Amber::Environment::Logger.new` because then you are calling default Crystal logger but with a couple of new methods added (`.info`, `.debug`, etc).

Instead, what you want to do is to call `LoggingBuilder`, the factory that creates a new `Logger` with Amber defaults.

TL;DR; `application.cr` is wrong, to keep the same format but specifying an output file it should be the following:

```ruby
  settings.logging.colorize = false
  settings.logger = Amber::Environment::LoggerBuilder.logger(File.open("./log/logs-tests.log", "w+"), settings.logging)
```

Also, as you can see, the factory accepts a `logging` parameter, and if you supply it with the application logging settings, they will be applied (in this case I am disabling colorize setting)

One thing that **isn't** working is logging options what accepts arrays, for example:

`settings.logging.filter = %w(password confirm_password)`

If you uncomment that settings you will get the following error:

`instance variable '@filter' of Amber::Environment::Logging must be Array(String | Nil), not Array(String)`

So, **I think** we should work towards fixing those bugs (logging settings that accepts arrays)

**Comment by docelic** [comment]
@epergo also there is one more improvement I thought of later -- currently you can call `logging.colorize = true/false`, but you can't call that on `logger`.

Maybe it would be good, simply for convenience/simplicity, to implement this method on `logger` too.

**Comment by danielRomero** [comment]
Hi,

I think that the logging is different than logger and each one has different responsabilities.

The `logger`'s responsability is where to output the info, meanwhile the `logging` responsability is to tell the `logger` how and what info must be logged, so I think the logger doesn't need to manage the configuration via `logger.colorize = true/false`

This is what I think about it and doesn't mean that it is correct, what do you think?

just my two cents!!

**Comment by docelic** [comment]
@danielRomero yes, my comment was in the sense that currently, I believe if you want to turn on or off coloring, you need to run `logging.colorize = false` and then replace the existing `logger` with a new one on which the colorize setting will be applied. So I was suggesting to have both the "template" (`logging`) and the actual logger (`logger`) to have this option for convenience.

**Comment by epergo** [comment]
I see, it can come in handy definitely. Do you think every option of logging should be 'proxied' to logger or only a subset?

Logging settings right now:
```
  # settings.logging.colorize = true
  # settings.logging.severity = debug
  # settings.logging.filter = %w(password confirm_password)
  # settings.logging.skip = %w()
  # settings.logging.context = %w(request headers cookies session params)
```

**Comment by docelic** [comment]
@epergo good question, if you look at item 2 from my initial text in this ticket, it mentions this:

"Some of the settings which exist on "logging" have non-standard names and values when compared to the Logger class.
For example, in "logging", the severity is called "severity" while in Logger it is called "level".
Also, severity (level) in "logging" is set as a lowercase string, while in Logger it is of type "Severity"."

So it seems to me that we should actually make all 'logging' and 'logger' settings uniform. (Logging should be modified to have the same names and types of arguments like Logger).

If we did that, then only a few methods would remain to be 'proxied', if at all.
(But yes, I would be in favor of making `logging` and `logger` accept the exact same arguments.)

You think that'd be a good approach? Also, what should the behavior be if someone modifies settings on `logger`? Should it be propagated to `logging` also, or it should be clear that `logging` only modifies the "template" for future loggers and `logger` only modifies the current/existing logger?

**Comment by epergo** [comment]
> For example, in "logging", the severity is called "severity" while in Logger it is called "level".
Also, severity (level) in "logging" is set as a lowercase string, while in Logger it is of type "Severity"."

Totally, should be called the same (`level`), but because logging is a 'user friendly' way to configure logging settings I think it should remain as a array of String, that way is easier to manage via yaml.

>  Also, what should the behavior be if someone modifies settings on logger? Should it be propagated to logging also, or it should be clear that logging only modifies the "template" for future loggers, and logger only modifies the current/existing logger?

I think logging and logger should remain separate, as you say, as a template for instantiating new loggers. And if you modify a logger 'in place' only that logger is modified.

-----

Having said that, I have the same feeling that you have (or I think you have) that logging settings are a bit confusing right as they are now. But I don't know how to improve it :sweat_smile:

**Comment by eliasjpr** [comment]
BTW we inherit the  Level == Severity from the Standard Lib see 

https://github.com/crystal-lang/crystal/blob/4f9ed8d03208dd0db33993c5a6fa6753bc1cf91e/src/logger.cr#L37

**Comment by eliasjpr** [comment]
Is this still an active issue? @amberframework/core-team

**Comment by drujensen** [comment]
Closing this issue. It seems to be working fine for me.
